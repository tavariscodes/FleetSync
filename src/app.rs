// Contains all of our application's state
// we will use application state to
// track the screen the user is seeing

use ratatui::crossterm::event::{self, Event, KeyCode, KeyEvent, KeyEventKind};
use ratatui::{Terminal, prelude::Backend};
use std::{default, fmt, io};

use crate::ui::ui;

#[derive(Debug, PartialEq)]
pub enum UserType {
    Driver,
    Dispatcher,
}

impl fmt::Display for UserType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            UserType::Driver => "Driver",
            UserType::Dispatcher => "Dispatcher",
        };
        write!(f, "{}", s)
    }
}
#[derive(Default, Debug)]
pub enum CurrentScreen {
    #[default]
    Main,
    Driver,
    Dispatch,
    Exiting,
}


#[derive(Debug, Default)]
pub struct App {
    pub user_type: Option<UserType>,
    pub current_screen: CurrentScreen,
    pub exit: bool,
}

impl App {
    pub fn new() -> App {
        App {
            user_type: None,
            current_screen: CurrentScreen::Main,
            exit: false,
        }
    }

    pub fn run<B: Backend>(&mut self, terminal: &mut Terminal<B>) -> io::Result<()> {
        while !self.exit {
            terminal.draw(|frame| ui(frame, self));
            self.handle_events()?;
        }
        Ok(())
    }

    pub fn handle_events(&mut self) -> io::Result<()> {
        match event::read()? {
            Event::Key(key_event) if key_event.kind == KeyEventKind::Press => {
                self.handle_key_event(key_event);
            }
            _ => {}
        };
        Ok(())
    }

    pub fn handle_exit_screen_key_event(&mut self, key_event: KeyEvent) {
        match key_event.code {
            KeyCode::Char('y') => {
                self.current_screen = CurrentScreen::Main;
                self.exit = true;
            }
            KeyCode::Char('n') => {
                self.current_screen = CurrentScreen::Main;
                self.user_type = None
            }
            _ => {}
        }
    }

    pub fn handle_driver_screen_key_event(&mut self, key_event: KeyEvent) {
        match key_event.code {
            KeyCode::Char('m') => {
                self.current_screen = CurrentScreen::Main;
                self.user_type = None;
            }
            _ => {}
        }
    }

    pub fn handle_dispatcher_screen_key_event(&mut self, key_event: KeyEvent) {
        match key_event.code {
            KeyCode::Char('m') => {
                self.current_screen = CurrentScreen::Main;
                self.user_type = None;
            }
            _ => {}
        }
    }

    pub fn handle_main_screen_key_event(&mut self, key_event: KeyEvent) {
        match key_event.code {
            KeyCode::Char('q') => {
                self.current_screen = CurrentScreen::Exiting;
            }
            KeyCode::Char('1') => {
                self.user_type = Some(UserType::Driver);
                self.current_screen = CurrentScreen::Driver;
            }
            KeyCode::Char('2') => {
                self.user_type = Some(UserType::Dispatcher);
                self.current_screen = CurrentScreen::Dispatch;
            }
            _ => {}
        }
    }

    pub fn handle_key_event(&mut self, key_event: KeyEvent) {
        match self.current_screen {
            CurrentScreen::Main => {
                self.handle_main_screen_key_event(key_event);
            }
            CurrentScreen::Driver => {
                self.handle_driver_screen_key_event(key_event);
            }
            CurrentScreen::Dispatch => {
                self.handle_dispatcher_screen_key_event(key_event);
            }
            CurrentScreen::Exiting => {
                self.handle_exit_screen_key_event(key_event);
            }
            _ => {}
        }
    }
}


#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use ratatui::{backend::TestBackend, Terminal};

    #[test]
    fn test_render_app() {
        let mut app = App::default();
    
        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal.draw(|frame| ui(frame, &app));
        
        assert_snapshot!(terminal.backend())
    }
}