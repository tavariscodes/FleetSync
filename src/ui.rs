use std::vec;

use ratatui::{
    Frame,
    layout::{Constraint, Direction, Layout},
    style::{Color, Style},
    text::{Line, Span, Text},
    widgets::{Block, Borders, Paragraph},
};

use crate::app::{App, CurrentScreen};

pub fn ui(frame: &mut Frame, app: &App) {
    let chunks = Layout::default()
        .direction(Direction::Vertical)
        .constraints([
            Constraint::Length(3),
            Constraint::Min(1),
            Constraint::Length(3),
        ])
        .split(frame.area());

    let title_block = Block::default()
        .borders(Borders::ALL)
        .style(Style::default());

    let title = Paragraph::new(Text::styled(
        "Welcome to FleetSync",
        Style::default().fg(Color::Green),
    ))
    .block(title_block);

    frame.render_widget(title, chunks[0]);

    let current_navigation_text = vec![
        // first half of text
        match app.current_screen {
            CurrentScreen::Main => Span::styled("Main Menu", Style::default().fg(Color::Green)),
            CurrentScreen::Dispatch => {
                Span::styled("Current Mode: Dispatcher", Style::default().fg(Color::Blue))
            }
            CurrentScreen::Driver => {
                Span::styled("Current Mode: Driver", Style::default().fg(Color::Blue))
            }
            CurrentScreen::Exiting => Span::styled("Exiting", Style::default().fg(Color::Blue)),
        }
        .to_owned(),
    ];

    let mode_footer = Paragraph::new(Line::from(current_navigation_text))
        .block(Block::default().borders(Borders::ALL));

    let current_keys_hint = {
        match app.current_screen {
            CurrentScreen::Main => Span::styled(
                "(q) to quit/ (1) login as driver / (2) login as dispatcher",
                Style::default().fg(Color::Red),
            ),
            CurrentScreen::Exiting => Span::styled(
                "(y)/(n) are you sure you want to quit?",
                Style::default().fg(Color::Red),
            ),
            _ => Span::styled("(m) back to main menu", Style::default().fg(Color::Red)),
        }
    };

    let key_notes_footer =
        Paragraph::new(Line::from(current_keys_hint)).block(Block::default().borders(Borders::ALL));

    let footer_chunks = Layout::default()
        .direction(Direction::Horizontal)
        .constraints([Constraint::Percentage(50), Constraint::Percentage(50)])
        .split(chunks[2]);

    frame.render_widget(mode_footer, footer_chunks[0]);
    frame.render_widget(key_notes_footer, footer_chunks[1]);
}
