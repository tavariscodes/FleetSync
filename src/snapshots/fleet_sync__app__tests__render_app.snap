---
source: src/app.rs
expression: terminal.backend()
---
"┌──────────────────────────────────────────────────────────────────────────────┐"
"│Welcome to FleetSync                                                          │"
"└──────────────────────────────────────────────────────────────────────────────┘"
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"┌──────────────────────────────────────┐┌──────────────────────────────────────┐"
"│Main Menu                             ││(q) to quit/ (1) login as driver / (2)│"
"└──────────────────────────────────────┘└──────────────────────────────────────┘"
